### DISCUSSION TITLE
Dynamically update matcher based on text!

### DISCUSSION THREAD

**Discussion by Ankush-Chander** [explosion/spaCy#11584] [Help: Coding & Implementations] [answered]
I have a use case where I need to apply matcher rules based on input text.

Steps look like:
1. Get Input text
2. Calculate candidate rules. (Here we rule out rules/patterns that are not relevant to text)
3. Apply candidate rules to text. 

Is it possible to dynamically add/remove/recreate patterns in  matcher based on the input text?

**Comment by polm** [comment, marked as answer]
This is possible, but it's kind of a weird thing to do. Usually you can just register all your patterns and ignore matches on patterns you don't want to use. For example, sometimes I make different Matcher objects and use them behind different branches (if statements) in code, but even that is rarely necessary. Can you be more specific about why you want to do this?

**Reply by Ankush-Chander** [reply]
So I have like 6k+rules/patterns that can be matched against an incoming sentence/text. Now I have implemented a filter such that given a sentence i can derive 5-10 rules that actually apply on that sentence(by indexing rules and matching sentence against it). 
so the steps  I have in mind are:
1. Calculate applicable rules(reduce rules to be applied from 6K to 5-10)
2. Then apply those rules/patterns on the sentence.

Using this I want to avoid every incoming sentence getting matched against 6K rules and potentially becoming a CPU bottleneck given that I am getting thousands of concurrent sentence requests.

**Reply by Ankush-Chander** [reply]
@ines @polm Is this right way to think of matchers?
Any pointers would be helpful.

**Reply by polm** [reply]
Please do not tag maintainers to get their attention.

Without knowing more about your input and rules it's not clear to me whether your approach will be useful or not. Have you tried just having one Matcher with all your patterns and seeing if it's slower than picking the right rules? For performance related problems, the best thing to do is always to actually measure a realistic workload and compare implementations.

**Reply by Ankush-Chander** [reply]
Understood. Thanks for your help.
